src/main/java/org/springframework/integration/aws/lock/DynamoDbLockRegistry.java

```diff
@@ -0,0 +1,552 @@
+		}
+
+		this.createTableLatch.countDown();
+	}
+
+	@Override
+	public void stop(Runnable callback) {
+		stop();
+		callback.run();
+	}
+
+	@Override
+	public void stop() {
+		this.running.set(false);
+	}
+
+	@Override
+	public boolean isRunning() {
+		return false;
+	}
+
+	@Override
+	public int getPhase() {
+		return Integer.MAX_VALUE;
+	}
+
+	private void awaitForActive() {
+		try {
+			this.createTableLatch.await(60, TimeUnit.SECONDS);
+		}
+		catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
+			throw new IllegalStateException("The DynamoDb table " + this.tableName +
+					" has not been created during " + 60 + " seconds");
+		}
+	}
+
+	@Override
+	public void destroy() throws Exception {
+		if (!this.executorExplicitlySet) {
+			((ExecutorService) this.executor).shutdown();
+		}
+
+		if (!this.dynamoDBLockClientExplicitlySet) {
+			this.dynamoDBLockClient.close();
+		}
+	}
+
+	@Override
+	public Lock obtain(Object lockKey) {
+		Assert.isInstanceOf(String.class, lockKey);
+		return this.locks.computeIfAbsent((String) lockKey, DynamoDbLock::new);
+	}
+
+	@Override
+	public void expireUnusedOlderThan(long age) {
+		Iterator<Map.Entry<String, DynamoDbLock>> iterator = this.locks.entrySet().iterator();
+		long now = System.currentTimeMillis();
+		while (iterator.hasNext()) {
+			Map.Entry<String, DynamoDbLock> entry = iterator.next();
+			DynamoDbLock lock = entry.getValue();
+			if (now - lock.lastUsed > age && !lock.delegate.isHeldByCurrentThread()) {
+				iterator.remove();
+			}
+		}
+	}
+
+	private final class DynamoDbLock implements Lock {
+
+		private final ReentrantLock delegate = new ReentrantLock();
+
+		private final String key;
+
+		// It is safe to use a shared instance - access is guaranteed by the delegate lock.
+		private final AcquireLockOptions.AcquireLockOptionsBuilder acquireLockOptionsBuilder;
+
+		private LockItem lockItem;
+
+		private volatile long lastUsed = System.currentTimeMillis();
+
+		private DynamoDbLock(String key) {
+			this.key = key;
+			this.acquireLockOptionsBuilder =
+					AcquireLockOptions.builder(this.key)
+							.withReplaceData(false)
+							.withSortKey(DynamoDbLockRegistry.this.sortKey)
+							.withTimeUnit(TimeUnit.MILLISECONDS)
+							.withRefreshPeriod(DynamoDbLockRegistry.this.refreshPeriod);
+		}
+
+		private void rethrowAsLockException(Exception e) {
+			throw new CannotAcquireLockException("Failed to lock at " + this.key, e);
+		}
+
+		@Override
+		public void lock() {
+			awaitForActive();
+
+			this.delegate.lock();
+
+			this.acquireLockOptionsBuilder
+					.withAdditionalTimeToWaitForLock(Long.MAX_VALUE - DynamoDbLockRegistry.this.leaseDuration);
+
+
+			try {
+				while (!doLock()) {
+					Thread.sleep(100); //NOSONAR
+				}
+			}
+			catch (InterruptedException e) {
+				/*
+				 * This method must be uninterruptible so catch and ignore
+				 * interrupts and only break out of the while loop when
+				 * we get the lock.
+				 */
```

Comment:
We should reinstate the interrupt before exiting, though - set a boolean and add finally to an outer try?

Reply:
That's just Dave Syer's magic copy/paste from the `JdbcLockRegistry` :smile: . Sorry, I don't understand what you mean.
Some code sample?

Reply:
```java
public void lock() {
	this.delegate.lock();
	boolean wasInterruptedWhileUninterruptible = false;
	try {
		while (true) {
			try {
				while (!doLock()) {
					Thread.sleep(100); //NOSONAR
				}
				break;
			}
			catch (CannotSerializeTransactionException | TransactionTimedOutException | QueryTimeoutException e) {
				// try again
			}
			catch (InterruptedException e) {
					/*
					 * This method must be uninterruptible so catch and ignore
					 * interrupts and only break out of the while loop when
					 * we get the lock.
					 */
				wasInterruptedWhileUninterruptible = true;
			}
			catch (Exception e) {
				this.delegate.unlock();
				rethrowAsLockException(e);
			}
		}
	}
	finally {
		if (wasInterruptedWhileUninterruptible) {
			Thread.currentThread().interrupt();
		}
	}
}
```